format/macho/macho.go

```diff
@@ -0,0 +1,492 @@
+	LC_IDFVMLIB:                 "LC_IDFVMLIB",
+	LC_IDENT:                    "LC_IDENT",
+	LC_FVMFILE:                  "LC_FVMFILE",
+	LC_PREPAGE:                  "LC_PREPAGE",
+	LC_DYSYMTAB:                 "LC_DYSYMTAB",
+	LC_LOAD_DYLIB:               "LC_LOAD_DYLIB",
+	LC_ID_DYLIB:                 "LC_ID_DYLIB",
+	LC_LOAD_DYLINKER:            "LC_LOAD_DYLINKER",
+	LC_ID_DYLINKER:              "LC_ID_DYLINKER",
+	LC_PREBOUND_DYLIB:           "LC_PREBOUND_DYLIB",
+	LC_ROUTINES:                 "LC_ROUTINES",
+	LC_SUB_FRAMEWORK:            "LC_SUB_FRAMEWORK",
+	LC_SUB_UMBRELLA:             "LC_SUB_UMBRELLA",
+	LC_SUB_CLIENT:               "LC_SUB_CLIENT",
+	LC_SUB_LIBRARY:              "LC_SUB_LIBRARY",
+	LC_TWOLEVEL_HINTS:           "LC_TWOLEVEL_HINTS",
+	LC_PREBIND_CKSUM:            "LC_PREBIND_CKSUM",
+	LC_LOAD_WEAK_DYLIB:          "LC_LOAD_WEAK_DYLIB",
+	LC_SEGMENT_64:               "LC_SEGMENT_64",
+	LC_ROUTINES_64:              "LC_ROUTINES_64",
+	LC_UUID:                     "LC_UUID",
+	LC_RPATH:                    "LC_RPATH",
+	LC_CODE_SIGNATURE:           "LC_CODE_SIGNATURE",
+	LC_SEGMENT_SPLIT_INFO:       "LC_SEGMENT_SPLIT_INFO",
+	LC_REEXPORT_DYLIB:           "LC_REEXPORT_DYLIB",
+	LC_LAZY_LOAD_DYLIB:          "LC_LAZY_LOAD_DYLIB",
+	LC_ENCRYPTION_INFO:          "LC_ENCRYPTION_INFO",
+	LC_DYLD_INFO:                "LC_DYLD_INFO",
+	LC_DYLD_INFO_ONLY:           "LC_DYLD_INFO_ONLY",
+	LC_LOAD_UPWARD_DYLIB:        "LC_LOAD_UPWARD_DYLIB",
+	LC_VERSION_MIN_MACOSX:       "LC_VERSION_MIN_MACOSX",
+	LC_VERSION_MIN_IPHONEOS:     "LC_VERSION_MIN_IPHONEOS",
+	LC_FUNCTION_STARTS:          "LC_FUNCTION_STARTS",
+	LC_DYLD_ENVIRONMENT:         "LC_DYLD_ENVIRONMENT",
+	LC_MAIN:                     "LC_MAIN",
+	LC_DATA_IN_CODE:             "LC_DATA_IN_CODE",
+	LC_SOURCE_VERSION:           "LC_SOURCE_VERSION",
+	LC_DYLIB_CODE_SIGN_DRS:      "LC_DYLIB_CODE_SIGN_DRS",
+	LC_ENCRYPTION_INFO_64:       "LC_ENCRYPTION_INFO_64",
+	LC_LINKER_OPTION:            "LC_LINKER_OPTION",
+	LC_LINKER_OPTIMIZATION_HINT: "LC_LINKER_OPTIMIZATION_HINT",
+	LC_VERSION_MIN_TVOS:         "LC_VERSION_MIN_TVOS",
+	LC_VERSION_MIN_WATCHOS:      "LC_VERSION_MIN_WATCHOS",
+	LC_NOTE:                     "LC_NOTE",
+	LC_BUILD_VERSION:            "LC_BUILD_VERSION",
+}
+
+func machoDecode(d *decode.D, in interface{}) interface{} {
+	var archBits int
+	var ncmds uint64
+	magicBuffer := d.U32LE()
+
+	if magicBuffer == MH_MAGIC || magicBuffer == MH_MAGIC_64 {
+		d.Endian = decode.LittleEndian
+		if magicBuffer == MH_MAGIC {
+			archBits = 32
+		} else {
+			archBits = 64
+		}
+	} else if magicBuffer == MH_CIGAM || magicBuffer == MH_CIGAM_64 {
+		d.Endian = decode.BigEndian
+		if magicBuffer == MH_CIGAM {
+			archBits = 32
+		} else {
+			archBits = 64
+		}
+	} else {
+		d.Fatalf("Invalid magic field")
+	}
+
+	d.SeekAbs(0)
+	d.FieldStruct(fmt.Sprintf("mach_header_%d", archBits), func(d *decode.D) {
+		d.FieldU32("magic", magicToHex, classBits, endianNames)
```

Comment:
I think you can replace `magicToHex` with `scalar.Hex`. 

Maybe can use `UToScalar` and use description? other alternatives is to use a struct or use `d.FieldValue*` if you want to add for a lack of better name "synthetic" values that don't have a bit range.

Reply:
The first paragraph is addressed.

I did not get what you meant by the second one.

Reply:
Sorry i should have been more clear. What i meant was that magic seems to both indicate endian and number of bits? so it wold be nice if both of those var available as separate fields. One way could be something like thus:
```go
// should maybe assign map to some global var
d.FieldU32("magic", scalar.Hex, scalar.UToScalar{
	MH_MAGIC:    scalar.S{Description: "32-bit little endian"},
	MH_CIGAM:    scalar.S{Description: "32-bit big endian"},
	MH_MAGIC_64: scalar.S{Description: "64-bit little endian"},
	MH_CIGAM_64: scalar.S{Description: "64-bit big endian"},
})
d.FieldValueU("bits", uint64(archBits))
d.FieldValueStr("endian", ...)
```

If you would like to build some kind of shared table for this you can do something similar to this https://github.com/wader/fq/blob/master/format/elf/elf.go#L251 then you can use the map as normal and implement scalar.Mapper or have methods on the type that return mappers. But maybe overkill?

Reply:
I saw this now and updated accordingly.